nrf52840-hal/src/ieee802154.rs

```diff
@@ -0,0 +1,630 @@
+        match event {
+            Event::End => {
+                while self.radio.events_end.read().events_end().bit_is_clear() {
+                    continue;
+                }
+                self.radio.events_end.reset();
+            }
+            Event::PhyEnd => {
+                while self
+                    .radio
+                    .events_phyend
+                    .read()
+                    .events_phyend()
+                    .bit_is_clear()
+                {
+                    continue;
+                }
+                self.radio.events_phyend.reset();
+            }
+        }
+    }
+
+    /// Waits until the radio state matches the given `state`
+    fn wait_for_state(&self, state: STATE_A) {
+        while self.radio.state.read().state() != state {
+            continue;
+        }
+    }
+}
+
+/// NOTE must be followed by a volatile write operation
+fn dma_start_fence() {
+    atomic::compiler_fence(Ordering::Release);
+}
+
+/// NOTE must be preceded by a volatile read operation
+fn dma_end_fence() {
+    atomic::compiler_fence(Ordering::Acquire);
+}
+
+enum Event {
+    End,
+    PhyEnd,
+}
+
+/// An IEEE 802.15.4 packet
+///
+/// This `Packet` is closest to the PPDU (PHY Protocol Data Unit) defined in the IEEE spec. The API
+/// lets users modify the payload of the PPDU via the `deref` and `copy_from_slice` methods. End
+/// users should write a MPDU (MAC protocol data unit), starting with a MAC header (MHDR), in the
+/// PPDU payload to be IEEE compliant.
+///
+/// Note that the MAC level CRC, AKA the FCS (Frame Control Sequence), is fully computed in hardware
+/// so it doesn't need to be included in the packet's payload
+///
+/// See figure 119 in the Product Specification of the nRF52840 for more details
+pub struct Packet {
+    buffer: [u8; Self::SIZE],
+}
+
+// See figure 124 in nRF52840-PS
+impl Packet {
+    const PHY_HDR: usize = 0;
+    const DATA: RangeFrom<usize> = 1..;
+    const CRC: u8 = 2; // size of the CRC, which is *never* copied to / from RAM
+    const SIZE: usize = 1 /* PHY_HDR */ + Self::MAX_LEN as usize + 1 /* LQI */;
+
+    /// The maximum length of the packet's payload
+    pub const MAX_LEN: u8 = 125;
+
+    /// Returns an empty packet (length = 0)
+    // XXX I believe that be making this not `const` it is not possible to place a `Packet` in
+    // `.rodata` (modulo `#[link_section]` shenanigans) thus it is not necessary to check the
+    // address of packet in `Radio.{send,recv}` (EasyDMA can only operate on RAM addresses)
+    pub fn new() -> Self {
+        let mut packet = Self {
+            buffer: [0; Self::SIZE],
+        };
+        packet.set_len(0);
+        packet
+    }
+
+    /// Fills the packet with given `src` data
+    ///
+    /// NOTE `src` data will be truncated to `MAX_PACKET_SIZE` bytes
+    pub fn copy_from_slice(&mut self, src: &[u8]) {
+        let len = cmp::min(src.len(), Self::MAX_LEN as usize) as u8;
```

Comment:
I don't think it is a good idea to silently limit the given slice size. When user of this function wants to transmit 200 bytes packet and calls this function, would get only 125 bytes transmitted and would not be notified that there was a problem. I would change function signature to return `Result` or add assertion.